File: C++/ps_baekjoon/10868/main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int INF = numeric_limits<int>::max() / 2;
int n, m;

class SegTree
{
private:
    vector<int> tree;
    const vector<int>& data;
public:
    SegTree(const vector<int>& data) :data{data} { tree.resize(data.size() * 4); }

    int merge(int left, int right) { return min(left, right); }

    int build(int node, int nodeLeft, int nodeRight)
    {
        if (nodeLeft == nodeRight)
            return tree[node] = data[nodeLeft];
        int mid = (nodeLeft + nodeRight) / 2;
        int leftVal = build(node * 2, nodeLeft, mid);
        int rightVal = build(node * 2 + 1, mid + 1, nodeRight);
        return tree[node] = merge(leftVal, rightVal);
    }

    int query(int left, int right, int node, int nodeLeft, int nodeRight)
    {
        if (right < nodeLeft || nodeRight < left)
            return INF;
        if (left <= nodeLeft && nodeRight <= right)
            return tree[node];
        int mid = (nodeLeft + nodeRight) / 2;
        int leftVal = query(left, right, node * 2, nodeLeft, mid);
        int rightVal = query(left, right, node * 2 + 1, mid + 1, nodeRight);
        return merge(leftVal, rightVal);
    }
};

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> m;
    vector<int> data(n + 1);
    for (int i = 1; i <= n; ++i)
        cin >> data[i];
    SegTree st(data);
    st.build(1, 1, n);
    while (m--)
    {
        int a, b;
        cin >> a >> b;
        cout << st.query(a, b, 1, 1, n) << '\n';
    }
    return 0;
}

File: Python/ps_baekjoon/10868/main.py
import sys

readline, write = (sys.stdin.readline, sys.stdout.write)
INF = float("inf")
res = []

n, m = [int(i) for i in readline().split()]


class SegTree:
    def __init__(self, data):
        self.data = data
        self.tree = [0] * (len(self.data) * 4)

    def merge(self, left, right):
        return min(left, right)

    def build(self, node=1, nodeLeft=1, nodeRight=n):
        if nodeLeft == nodeRight:
            self.tree[node] = self.data[nodeLeft]
            return self.tree[node]
        mid = (nodeLeft + nodeRight) // 2
        leftVal = self.build(node * 2, nodeLeft, mid)
        rightVal = self.build(node * 2 + 1, mid + 1, nodeRight)
        self.tree[node] = self.merge(leftVal, rightVal)
        return self.tree[node]

    def query(self, left, right, node=1, nodeLeft=1, nodeRight=n):
        if right < nodeLeft or nodeRight < left:
            return INF
        if left <= nodeLeft and nodeRight <= right:
            return self.tree[node]
        mid = (nodeLeft + nodeRight) // 2
        leftVal = self.query(left, right, node * 2, nodeLeft, mid)
        rightVal = self.query(left, right, node * 2 + 1, mid + 1, nodeRight)
        return self.merge(leftVal, rightVal)


data = [0] + [[int(i) for i in readline().split()][0] for _ in range(n)]
st = SegTree(data)
st.build()
for _ in range(m):
    a, b = [int(i) for i in readline().split()]
    res.append(str(st.query(a, b)))
write("\n".join(res))